manuals/luksdeinfo.1
.Dd June 26, 2026
.Dt LUKSDEINFO 1
.Os
.Sh NAME
.Nm luksdeinfo
.Nd determines information about a Linux Unified Key Setup (LUKS) volume
.Sh SYNOPSIS
.Nm luksdeinfo
.Op Fl k Ar key
.Op Fl o Ar offset
.Op Fl p Ar password
.Op Fl huvV
.Ar source
.Sh DESCRIPTION
.Nm luksdeinfo
is a utility to determine information about a Linux Unified Key Setup (LUKS) \
volume
.Pp
.Nm luksdeinfo
is part of the
.Nm libluksde
package.
.Nm libluksde
is a library to access the Linux Unified Key Setup (LUKS) Disk Encryption format
.Pp
.Ar source
is the source volume.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl h
shows this help
.It Fl k Ar key
specify the key formatted in base16
.It Fl o Ar offset
specify the volume offset in bytes
.It Fl p Ar password
specify the password (or passphrase)
.It Fl u
unattended mode (disables user interaction)
.It Fl v
verbose output to stderr
.It Fl V
print version
.El
.Sh ENVIRONMENT
None
.Sh FILES
None
.Sh EXAMPLES
.Bd -literal
# luksdeinfo -p Password /dev/sda1
luksdeinfo 20130704
.sp
Linux Unified Key Setup Disk Encryption information:
	Encryption method:		AES 128-bit
.sp
.Ed
.Sh DIAGNOSTICS
Errors, verbose and debug output are printed to stderr when verbose output \
\-v is enabled.
Verbose and debug output are only printed when enabled at compilation.
.Sh SEE ALSO
.Xr luksdemount 1
.Sh AUTHORS
.An Joachim Metz <joachim.metz@gmail.com>
.Sh BUGS
Please report bugs of any kind on the project issue tracker: \
https://github.com/libyal/libluksde/issues
.Sh COPYRIGHT
Copyright (C) 2013-2026, Joachim Metz <joachim.metz@gmail.com>.
.sp
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A \
PARTICULAR PURPOSE.


ChangeLog
TODO
* add more password hashing support
  - fix caluclation of validation hash for SHA-224 and SHA-256 password hash
* documentation
  - add SHA224 hashing method
* add thread-safety in volume
  - add tests
* detect encryption types (notify as unsupported)
* add libluksde_diffuser_split?

* library: add support to set the master key directly

Crypto support:
* add support for lmk
* add support for serpent-CBC
* add support for twofish
* add support for cast5
* add support for cast6

Hashing support:
* add support for RIPEMD160
* add support for whirlpool

20141001
* see `git log' for more recent change log
* removed README.macosx
* changes for project site move

20140801
* bug fix in Python-bindings

20140531
* updated dependencies

20140330
* worked on Python bindings

20140112
* 2014 update
* updated dependencies

20131205
* updated dependencies

20130909
* updated dependencies
* worked on libcthreads build support

20130716
* updated dependencies

20130714
* worked on SHA512 support

20130704
* worked on initial version
* renamed libluksve to libluksde
* removed unused local libraries

20130703
* initial version based on libbde



luksdetools/luksdeinfo.c
/*
 * Shows information obtained from a Linux Unified Key Setup (LUKS) volume.
 *
 * Copyright (C) 2013-2026, Joachim Metz <joachim.metz@gmail.com>
 *
 * Refer to AUTHORS for acknowledgements.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include <common.h>
#include <file_stream.h>
#include <system_string.h>
#include <types.h>

#if defined( HAVE_FCNTL_H ) || defined( WINAPI )
#include <fcntl.h>
#endif

#if defined( HAVE_IO_H ) || defined( WINAPI )
#include <io.h>
#endif

#if defined( HAVE_STDLIB_H ) || defined( WINAPI )
#include <stdlib.h>
#endif

#if defined( HAVE_UNISTD_H )
#include <unistd.h>
#endif

#include "info_handle.h"
#include "luksdetools_getopt.h"
#include "luksdetools_libcerror.h"
#include "luksdetools_libclocale.h"
#include "luksdetools_libcnotify.h"
#include "luksdetools_libluksde.h"
#include "luksdetools_output.h"
#include "luksdetools_signal.h"
#include "luksdetools_unused.h"

info_handle_t *luksdeinfo_info_handle = NULL;
int luksdeinfo_abort                  = 0;

/* Signal handler for luksdeinfo
 */
void luksdeinfo_signal_handler(
      luksdetools_signal_t signal LUKSDETOOLS_ATTRIBUTE_UNUSED )
{
	libcerror_error_t *error = NULL;
	static char *function    = "luksdeinfo_signal_handler";

	LUKSDETOOLS_UNREFERENCED_PARAMETER( signal )

	luksdeinfo_abort = 1;

	if( luksdeinfo_info_handle != NULL )
	{
		if( info_handle_signal_abort(
		     luksdeinfo_info_handle,
		     &error ) != 1 )
		{
			libcnotify_printf(
			 "%s: unable to signal info handle to abort.\n",
			 function );

			libcnotify_print_error_backtrace(
			 error );
			libcerror_error_free(
			 &error );
		}
	}
	/* Force stdin to close otherwise any function reading it will remain blocked
	 */
#if defined( WINAPI ) && !defined( __CYGWIN__ )
	if( _close(
	     0 ) != 0 )
#else
	if( close(
	     0 ) != 0 )
#endif
	{
		libcnotify_printf(
		 "%s: unable to close stdin.\n",
		 function );
	}
}

/* The main program
 */
#if defined( HAVE_WIDE_SYSTEM_CHARACTER )
int wmain( int argc, wchar_t * const argv[] )
#else
int main( int argc, char * const argv[] )
#endif
{
	const char *description = \
		"Use luksdeinfo to determine information about a Linux Unified Key Setup (LUKS) volume.";

	luksdetools_option_t options[ ] = {
		{ 'h', NULL, "shows this help" },
		{ 'k', "key", "specify the key formatted in base16" },
		{ 'o', "offset", "specify the volume offset in bytes" },
		{ 'p', "password", "specify the password (or passphrase)" },
		{ 'u', NULL, "unattended mode (disables user interaction)" },
		{ 'v', NULL, "verbose output to stderr" },
		{ 'V', NULL, "print version" },
		{ 0, "source", "the source volume" },
	};
	system_character_t options_string[ 32 ];

	libluksde_error_t *error            = NULL;
	system_character_t *option_key      = NULL;
	system_character_t *option_offset   = NULL;
	system_character_t *option_password = NULL;
	system_character_t *source          = NULL;
	char *program                       = "luksdeinfo";
	system_integer_t option             = 0;
	int number_of_options               = (int) ( sizeof( options ) / sizeof( luksdetools_option_t ) );
	int unattended_mode                 = 0;
	int verbose                         = 0;

#if defined( __MINGW32__ ) && defined( HAVE_MINGW_BINMODE )
	_setmode( _fileno( stdout ), _O_BINARY );
	_setmode( _fileno( stderr ), _O_BINARY );
#endif

	libcnotify_stream_set(
	 stderr,
	 NULL );
	libcnotify_verbose_set(
	 1 );

	if( libclocale_initialize(
	     "luksdetools",
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize locale values.\n" );

		goto on_error;
	}
	if( luksdetools_output_initialize(
	     _IONBF,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize output settings.\n" );

		goto on_error;
	}
	luksdetools_output_version_fprint(
	 stdout,
	 program );

	if( luksdetools_getopt_get_options_string(
	     options,
	     number_of_options,
	     options_string,
	     32 ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to determine options string.\n" );

		goto on_error;
	}
	while( ( option = luksdetools_getopt(
	                   argc,
	                   argv,
	                   options_string ) ) != (system_integer_t) -1 )
	{
		switch( option )
		{
			case (system_integer_t) '?':
			default:
				fprintf(
				 stderr,
				 "Invalid argument: %" PRIs_SYSTEM "\n",
				 argv[ optind - 1 ] );

				luksdetools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_FAILURE );

			case (system_integer_t) 'h':
				luksdetools_getopt_usage_fprint(
				 stdout,
				 program,
				 description,
				 options,
				 number_of_options );

				return( EXIT_SUCCESS );

			case (system_integer_t) 'k':
				option_key = optarg;

				break;

			case (system_integer_t) 'o':
				option_offset = optarg;

				break;

			case (system_integer_t) 'p':
				option_password = optarg;

				break;

			case (system_integer_t) 'u':
				unattended_mode = 1;

				break;

			case (system_integer_t) 'v':
				verbose = 1;

				break;

			case (system_integer_t) 'V':
				luksdetools_output_copyright_fprint(
				 stdout );

				return( EXIT_SUCCESS );
		}
	}
	if( optind == argc )
	{
		fprintf(
		 stderr,
		 "Missing source volume.\n" );

		luksdetools_getopt_usage_fprint(
		 stdout,
		 program,
		 description,
		 options,
		 number_of_options );

		return( EXIT_FAILURE );
	}
	source = argv[ optind ];

	libcnotify_verbose_set(
	 verbose );
	libluksde_notify_set_stream(
	 stderr,
	 NULL );
	libluksde_notify_set_verbose(
	 verbose );

	if( info_handle_initialize(
	     &luksdeinfo_info_handle,
	     unattended_mode,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to initialize info handle.\n" );

		goto on_error;
	}
	if( option_key != NULL )
	{
		if( info_handle_set_key(
		     luksdeinfo_info_handle,
		     option_key,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set key.\n" );

			goto on_error;
		}
	}
	if( option_password != NULL )
	{
		if( info_handle_set_password(
		     luksdeinfo_info_handle,
		     option_password,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set password.\n" );

			goto on_error;
		}
	}
	if( option_offset != NULL )
	{
		if( info_handle_set_volume_offset(
		     luksdeinfo_info_handle,
		     option_offset,
		     &error ) != 1 )
		{
			fprintf(
			 stderr,
			 "Unable to set volume offset.\n" );

			goto on_error;
		}
	}
	if( info_handle_open(
	     luksdeinfo_info_handle,
	     source,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to open source volume.\n" );

		goto on_error;
	}
	if( info_handle_volume_fprint(
	     luksdeinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to print volume information.\n" );

		goto on_error;
	}
	if( info_handle_close(
	     luksdeinfo_info_handle,
	     &error ) != 0 )
	{
		fprintf(
		 stderr,
		 "Unable to close info handle.\n" );

		goto on_error;
	}
	if( info_handle_free(
	     &luksdeinfo_info_handle,
	     &error ) != 1 )
	{
		fprintf(
		 stderr,
		 "Unable to free info handle.\n" );

		goto on_error;
	}
	return( EXIT_SUCCESS );

on_error:
	if( error != NULL )
	{
		libcnotify_print_error_backtrace(
		 error );
		libcerror_error_free(
		 &error );
	}
	if( luksdeinfo_info_handle != NULL )
	{
		info_handle_free(
		 &luksdeinfo_info_handle,
		 NULL );
	}
	return( EXIT_FAILURE );
}

